### PR TITLE
Playwright e2e

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

In order to get your environment set up to test this code, run [e2e/playwright/setup-environment](https://github.com/mickmister/mattermost-plugin-github/blob/e2e-playwright/e2e/playwright/setup-environment) or run equivalent commands for your current setup.

What the script does:

- Navigate to the folder above `mattermost-plugin-github`
- Clone `mattermost-server` (if it is already cloned there, please have a clean git index to avoid issues with conflicts)
- `cd mattermost-server`
- Install webapp dependencies - `cd webapp && npm i`
- Install Playwright test dependencies - `cd ../e2e/playwright && npm i`
- Install Playwright - `npx install playwright`
- Install GitHub plugin e2e dependencies - `cd ../../../mattermost-plugin-github/e2e/playwright && npm i`
- Build and deploy plugin with mock OAuth server URL env var - `PLUGIN_E2E_MOCK_OAUTH_SERVER_URL=http://localhost:8080 make deploy`

-----

Then to run the tests:

Start Mattermost server:
- `cd <path>/mattermost-server/server`
- `make run-server`
- `make test-data`

Run test:
- Create a personal access token from GitHub
- Set `PLUGIN_E2E_MOCK_OAUTH_TOKEN` environment variable to access token
- `cd <path>/mattermost-plugin-github/e2e/playwright`
- `npm test`

To see the test report:
- `cd <path>/mattermost-plugin-github/e2e/playwright`
- `npm run show-report`
- Navigate to http://localhost:9323

To see test screenshots:
- `cd <path>/mattermost-plugin-github/e2e/playwright/screenshots`

---

I've added the following environment variables for different pieces of the system. The ones marked as "required" must be set in order for the setup to work properly:

- GitHub plugin compilation
    - required: `PLUGIN_E2E_MOCK_OAUTH_SERVER_URL` = http://localhost:8080
- OAuth server settings
    - required: `PLUGIN_E2E_MOCK_OAUTH_TOKEN` - Hardcoded token to use for OAuth handshake
    - optional: `NODE_ENV`='development' to log requests/responses
    - optional: `MM_SITE_URL` || 'http://localhost:8065'
    - optional: `MM_PLUGIN_ID` || 'github'
    - optional: `OAUTH_SERVER_PORT` || 8080;
- Playwright
    - optional: `MM_DATABASE_CONNECTION_STRING` || 'postgres://mmuser:mostest@localhost/mattermost_test?sslmode=disable\u0026connect_timeout=10\u0026binary_parameters=yes';

### CODE REVIEW

**Comment by mickmister** [comment]
The idea of the `mattermost-plugin-test-utils` folder is to be a git submodule in each of the plugin repos. This way the tooling is defined in a separate project and controlled by git commit. CI is then free to change this version at runtime.

File: e2e/playwright/package.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "plugin-e2e-tests",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "PLAYWRIGHT_TEST_DIR='../../../mattermost-plugin-github/e2e/playwright' npm run test-slomo --prefix ../../../mattermost-webapp/e2e/playwright"
```
Comment (mickmister) [inline e2e/playwright/package.json:7]: Accessing mm-webapp in parent directory

File: e2e/playwright/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+      "strictNullChecks": true,
+      "skipLibCheck": true,
+      "noEmit": true,
+      "baseUrl": ".",
+      "paths": {
+          "@mattermost/client/*": ["../../../mattermost-webapp/packages/client/lib/*"],
+          "@mattermost/types/*": ["../../../mattermost-webapp/packages/types/lib/*"],
+          "@e2e-support/*": ["../../../mattermost-webapp/e2e/playwright/support/*"],
+          "@e2e-test.config": ["../../../mattermost-webapp/e2e/playwright/test.config.ts"],
+      }
```
Comment (mickmister) [inline e2e/playwright/tsconfig.json:14]: Accessing mm-webapp in parent directory here too

**Review by m1lt0n — COMMENTED** [review]
I really like the approach @mickmister ! Great stuff! I guess there is some follow up work on this (e.g. CI, also the pending discussion on the dependencies for webapp (which though I think is non-blocking)), or can this be merged and then build on top of it (asking because I see some additional commits after you created the PR)?

**Comment by mickmister** [comment]
An improvement here would be to make the OAuth server run automatically within the test setup itself

**Pull request closed without merging**